### ISSUE TITLE
STRIDE parameter for FIT function

### ISSUE THREAD

**Issue by korosig** [unit8co/darts#754]
The backtest and the historical_forecasts function include a STRIDE parameter, but how can I train a model with the same data samples without a STRIDE parameter.

What if I want to move forward not just one step in my data, but several steps?  
For example, I have a huge (hourly based) data set and every morning at 8am I want to predict something for the next 24 hours? 
The way I see it, currently (without STRIDE) I have to run the model for every hour and it will find the best solution (with the least error) to do this. However, it is irrelevant to me how accurately the model predicts at 5pm for the next 24 hours, as I am only interested in the 8am hour. 

Without Stride, I can train the other hours without any need, and the program makes a very good model, which may be very accurate for all hours but may create huge error for my observed hour.

I could give you a few more examples. For example, we always want Monday to predict something for next week, etc.

**Comment by korosig** [comment]
![image](https://user-images.githubusercontent.com/23563516/150917736-1a1fd1b4-a277-473b-9ec4-fab8346cb971.png)

**Comment by brunnedu** [comment]
Hi @korosig, the `historical_forecasts` method is useful for testing how your model would've performed if it had been used historically. By setting the `stride=24`, the forecast horizon `n=24`, `last_points_only=False` and the start to a timestamp on a day after the start of your training dataset (say January 1st 2000) at 8am `start=pd.Timestamp('2000-01-01 08:00:00')`, the historical forecasts method will first fit the model with all data up to `start` and predict the next `n=24` hourly time steps. It will then move `stride=24` time steps ahead and again fit the model with all the data up to that point and again predict the next `n=24` hourly timesteps. This will continue until the last time step at 8am in your training dataset. The result is a contiguous prediction time series that is probably most accurate right after 8am on a given day and then loses accuracy until 8am the next day. If you'd like to understand the other parameters of the historical forecasting method feel free to check out the historical_forecasts method in the [docs](https://unit8co.github.io/darts/generated_api/darts.models.forecasting.baselines.html#darts.models.forecasting.baselines.NaiveDrift.historical_forecasts), it is the same for all forecasting models.

**Comment by korosig** [comment]
So it is working like this? 
1. iteration
                         history                              | prediction
    n past steps           train(24 steps)         | 24 steps
>>> .......... >>> ++++++++++++++++|
      n past steps 		                          |
>>>>>>>>>>>>>> ........................... >>>|***********
                                                                                  |
2. iteration                                                                 
					                                          |                  
                                  history                                     |  prediction
              n past steps                train(24 steps)          |  24 steps
>>>>>>>>>  .......... >>> ++++++++++++++++| 
      				n past steps                               |
>>>>>>>>>>>>>>>>>>>>  ........................... >>>|***********
                                                                                                  |
3. iteration
							                                          |
                          		         history                                      |  prediction
               		      n past steps            train(24 steps)              |  24 steps
>>>>>>>>>>>> >>> .......... >>> ++++++++++++++++ | 
      					n past steps 	                                  |
>>>>>>>>>>>>>>>>>>>>>>>>>>> ............................>>>|***********

**Comment by korosig** [comment]
So it is working like this? 
![image](https://user-images.githubusercontent.com/23563516/150948866-44db988e-f0ce-405d-875a-0aa8308b7610.png)

**Comment by brunnedu** [comment]
Yes, except that it trains on **all** the available past data and not just the previous 24 steps.

**Comment by hrzn** [comment]
I believe this was discussed on the Discord server. I'm pasting the discussion here:

-----------------------
Basically you would have to implement your own darts.utils.data.training_dataset.TrainingDataset . The flavour needs to depend on the kind of model you use. For instance for models supporting past covariates only, you would have to implement a PastCovariatesTrainingDataset , which you can see as a "generator" of tuples (past_target, past_covariate, future_target) each being a NumPy array. The datasets inherit from PyTorch Datasets. You will have to implement the __getitem__(i) function, which basically has to return the i -th slice of the target (and covariates). You can write this function such that for instance i=1 maps to a slice 24 positions after that of i=0 , etc.

You can see some docs here: https://unit8co.github.io/darts/generated_api/darts.utils.data.training_dataset.html
And an example implementation here: https://unit8co.github.io/darts/generated_api/darts.utils.data.sequential_dataset.html#darts.utils.data.sequential_dataset.PastCovariatesSequentialDataset (click on "source" to see the code)
-----------------------

@korosig if you implement a new kind of dataset supporting strides for training, please let us know; we might want to include a stride parameter in the fit() function.
I'll close this one for now.
